compiler_gym/util/minimize_trajectory.py

```diff
@@ -0,0 +1,317 @@
+    trajectory.
+
+    For each round of minimization, a number of actions are discarded randomly
+    and the hypothesis is tested. If the hypothesis still holds with those
+    actions removed, the minimization proceeds. Else the actions are re-inserted
+    into the trajectory and a new set of actions are removed. After a failure
+
+    Performs up to O(num_steps_ratio_multiplier * log n) evaluation rounds,
+    where n is the length of the trajectory.
+
+    :param env: An environment whose action trajectory should be minimized.
+    :param hypothesis: The hypothesis that is used to determine if a trajectory
+        is valid. A callback that accepts as argument the :code:`env:`
+        instance and returns true if the hypothesis holds, else false. The
+        hypothesis must hold on the initial trajectory.
+    :param num_steps_ratio_multiplier: A multiplier for the number of rounds of
+        minimization to perform, using log(n) the length of the trajectory as
+        the factor.
+    :param init_discard_ratio: The number of actions that will be randomly
+        discarded, as a multiplier of the length of the trajectory.
+    :param discard_ratio_decay: The ratio of decay for the discard ratio on
+        failure.
+    :param flakiness: The maximum number of times the hypothesis is repeated
+        to check if it holds.
+    :returns: A generator that yields the input environment every time the
+        trajectory is successfully reduced.
+    :raises MinimizationError: If the minimization fails, or if the hypothesis
+        does not hold on the initial trajectory.
+    """
+
+    def apply_and_test(actions):
+        return _apply_and_test(env, actions, hypothesis, flakiness)
+
+    actions = env.actions.copy()
+    if not apply_and_test(actions):
+        raise ValueError(
+            "Hypothesis failed on the initial state! The hypothesis must hold for the first state."
+        )
+
+    max_num_steps = int(log(len(actions), 2) * num_steps_ratio_multiplier)
+
+    num_steps = 0
+    discard_ratio = init_discard_ratio
+    while len(actions) > 5 and num_steps < max_num_steps:
+        num_steps += 1
+        num_to_remove = int(ceil(len(actions) * discard_ratio))
+        candidate_actions = actions.copy()
+        for _ in range(num_to_remove):
+            candidate_actions.remove(random.choice(candidate_actions))
+        if apply_and_test(candidate_actions):
+            logging.info(
+                "🟢 Hypothesis holds with %s of %s actions randomly removed, continuing",
+                num_to_remove,
+                len(actions),
+            )
+            actions = candidate_actions
+            discard_ratio = init_discard_ratio
+            yield env
+        else:
+            logging.info(
+                "🔴 Hypothesis does not hold with %s of %s actions randomly removed, rolling back",
+                num_to_remove,
+                len(actions),
+            )
+            discard_ratio *= discard_ratio_decay
```

Comment:
I feel like I could just be unlucky. Can't I have some number of attempts at a given `discard_ratio`?

Reply:
The reason for only using a single attempt at each discard ratio is because I found that being super aggressive with the ratio (75%) and backing it off quickly was more effective than being more conservative and trying for more attempts. I might be dead wrong about here, it was just a hunch I picked up while playing around with a few different buggy trajectories that I was minimizing :)